presto-hive/src/main/java/com/facebook/presto/hive/HivePartitionManager.java

```diff
@@ -143,16 +162,41 @@ public Iterable<HivePartition> getPartitionsIterator(
         else {
             return () -> {
                 List<String> partitionNames = partitionFilteringFromMetastoreEnabled ? getFilteredPartitionNames(session, metastore, hiveTableHandle, effectivePredicate) : getAllPartitionNames(session, metastore, hiveTableHandle, constraint);
-                return partitionNames.stream()
-                        // Apply extra filters which could not be done by getFilteredPartitionNames
-                        .map(partitionName -> parseValuesAndFilterPartition(tableName, partitionName, partitionColumns, partitionTypes, constraint))
-                        .filter(Optional::isPresent)
-                        .map(Optional::get)
-                        .iterator();
+
+                if (isParallelPartitionValueParsingEnabled(session) && partitionNames.size() > PARTITION_NAMES_BATCH_SIZE) {
+                    List<List<String>> partitionBatches = Lists.partition(partitionNames, PARTITION_NAMES_BATCH_SIZE);
+                    // Use CopyOnWriteArrayList to prevent race condition when multiple threads try to add partitions to this list
+                    List<HivePartition> result = new CopyOnWriteArrayList<>();
+                    List<ListenableFuture<?>> futures = new ArrayList<>();
+                    try {
+                        partitionBatches.forEach(batch -> futures.add(executorService.submit(() -> result.addAll(getPartitionListFromPartitionNames(batch, tableName, partitionColumns, partitionTypes, constraint)))));
+                        Futures.transform(Futures.allAsList(futures), input -> result, directExecutor()).get();
+                        return result.iterator();
+                    }
+                    catch (InterruptedException | ExecutionException e) {
+                        log.error("Parallel partition value parsing failed with error message:%s", e.getMessage());
```

Comment:
Chain the exception properly please.

Reply:
Can you please share some screen shots of number of threads it created when bombarded with queries?